File: MercuryAPI-1.31.3.36/java/samples_nb/src/samples/filter.java
/**
 * Sample program that demonstrates different types and uses of
 * TagFilter objects.
 */

// Import the API
package samples;
import com.thingmagic.*;

public class filter
{
  static SerialPrinter serialPrinter;
  static StringPrinter stringPrinter;
  static TransportListener currentListener;

  static void usage()
  {
    System.out.printf("Usage: Please provide valid arguments, such as:\n"
                + "filter [-v] [reader-uri] [--ant n[,n...]] \n" +
                  "-v  Verbose: Turn on transport listener\n" +
                  "reader-uri  Reader URI: e.g., \"tmr:///COM1\", \"tmr://astra-2100d3\"\n"
                + "--ant  Antenna List: e.g., \"--ant 1\", \"--ant 1,2\"\n"
                + "e.g: tmr:///com1 --ant 1,2 ; tmr://10.11.115.32 --ant 1,2\n ");
    System.exit(1);
  }

   public static void setTrace(Reader r, String args[])
  {
    if (args[0].toLowerCase().equals("on"))
    {
        r.addTransportListener(Reader.simpleTransportListener);
        currentListener = Reader.simpleTransportListener;
    }
    else if (currentListener != null)
    {
        r.removeTransportListener(Reader.simpleTransportListener);
    }
  }

   static class SerialPrinter implements TransportListener
  {
    public void message(boolean tx, byte[] data, int timeout)
    {
      System.out.print(tx ? "Sending: " : "Received:");
      for (int i = 0; i < data.length; i++)
      {
        if (i > 0 && (i & 15) == 0)
          System.out.printf("\n         ");
        System.out.printf(" %02x", data[i]);
      }
      System.out.printf("\n");
    }
  }

  static class StringPrinter implements TransportListener
  {
    public void message(boolean tx, byte[] data, int timeout)
    {
      System.out.println((tx ? "Sending:\n" : "Receiving:\n") +
                         new String(data));
    }
  }
  public static void main(String argv[])
  {
    // Program setup
    Reader r = null;
    int nextarg = 0;
    boolean trace = false;
    int[] antennaList = null;

    if (argv.length < 1)
      usage();

    if (argv[nextarg].equals("-v"))
    {
      trace = true;
      nextarg++;
    }

    // Create Reader object, connecting to physical device
    try
    {

        TagReadData[] tagReads, filteredTagReads;
        TagFilter filter;
        String readerURI = argv[nextarg];
        nextarg++;

        for (; nextarg < argv.length; nextarg++)
        {
            String arg = argv[nextarg];
            if (arg.equalsIgnoreCase("--ant"))
            {
                if (antennaList != null)
                {
                    System.out.println("Duplicate argument: --ant specified more than once");
                    usage();
                }
                antennaList = parseAntennaList(argv, nextarg);
                nextarg++;
            }
            else
            {
                System.out.println("Argument " + argv[nextarg] + " is not recognised");
                usage();
            }
        }

        r = Reader.create(readerURI);
        if (trace)
        {
          setTrace(r, new String[] {"on"});
        }
        
        r.connect();
        
        if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet("/reader/region/id"))
        {
            Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
            if (supportedRegions.length < 1)
            {
                throw new Exception("Reader doesn't support any regions");
            }
            else
            {
                r.paramSet("/reader/region/id", supportedRegions[0]);
            }
        }

        if (r.isAntDetectEnabled(antennaList))
        {
            System.out.println("Module doesn't has antenna detection support, please provide antenna list");
            r.destroy();
            usage();
        }

        SimpleReadPlan plan = new SimpleReadPlan(antennaList, TagProtocol.GEN2, null, null, 1000);
        r.paramSet(TMConstants.TMR_PARAM_READ_PLAN, plan);
        //Use first antenna for tag operation
        if (antennaList != null)
        {
            r.paramSet("/reader/tagop/antenna", antennaList[0]);
        }

      // In the current system, sequences of Gen2 operations require Session 0,
        // since each operation resingulates the tag.  In other sessions,
        // the tag will still be "asleep" from the preceding singulation.
        Gen2.Session oldSession = (Gen2.Session) r.paramGet("/reader/gen2/session");
        Gen2.Session newSession = Gen2.Session.S0;
        System.out.println("Changing to Session " + newSession + " (from Session " + oldSession + ")");
        r.paramSet("/reader/gen2/session", newSession);
        System.out.println();

        try
        {
            System.out.println("Unfiltered Read:");
            // Read the tags in the field
            tagReads = r.read(500);
            for (TagReadData tr : tagReads)
            {
                System.out.println(tr.toString());
            }
            System.out.println();

            if (0 == tagReads.length)
            {
                System.out.println("No tags found.");
            }
            else
            {
                // A TagData object may be used as a filter, for example to
                // perform a tag data operation on a particular tag.
                System.out.println("Filtered Tagop:");
                // Read kill password of tag found in previous operation
                filter = tagReads[0].getTag();
                System.out.printf("Read kill password of tag %s\n", filter);
                Gen2.ReadData tagop = new Gen2.ReadData(Gen2.Bank.RESERVED, 0, (byte) 2);
                try
                {
                    short[] data = (short[]) r.executeTagOp(tagop, filter);
                    for (short word : data)
                    {
                        System.out.printf("%04X", word);
                    }
                    System.out.println();
                }
                catch (ReaderException re)
                {
                    System.out.printf("Can't read tag: %s\n", re);
                }
                System.out.println();

                // Filter objects that apply to multiple tags are most useful in
                // narrowing the set of tags that will be read. This is
                // performed by setting a read plan that contains a filter.
                // A TagData with a short EPC will filter for tags whose EPC
                // starts with the same sequence.
                filter = new TagData(tagReads[0].getTag().epcString().substring(0, 4));
                System.out.printf("EPCs that begin with %s\n", filter);
                r.paramSet("/reader/read/plan",
                        new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, 1000));
                filteredTagReads = r.read(500);
                for (TagReadData tr : filteredTagReads)
                {
                    System.out.println(tr.toString());
                }
                System.out.println();

                // A filter can also be a full Gen2 Select operation.  For
                // example, this filter matches all Gen2 tags where bits 8-19 of
                // the TID are 0x003 (that is, tags manufactured by Alien
                // Technology). In case of Network readers, ensure that bitLength is
                // a multiple of 8.
                System.out.println("Tags with Alien Technology TID");
                filter = new Gen2.Select(false, Gen2.Bank.TID, 8, 12, new byte[] {0, 0x30});
                r.paramSet("/reader/read/plan",
                        new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, 1000));
                System.out.printf("Reading tags with a TID manufacturer of 0x003\n",
                        filter.toString());
                filteredTagReads = r.read(500);
                for (TagReadData tr : filteredTagReads)
                {
                    System.out.println(tr.toString());
                }
                System.out.println();

                // Gen2 Select may also be inverted, to give all non-matching tags
                // In case of Network readers, ensure that bitLength is a multiple of 8.
                System.out.println("Tags without Alien Technology TID");
                filter = new Gen2.Select(true, Gen2.Bank.TID, 8, 12, new byte[] {0, 0x30});
                r.paramSet("/reader/read/plan",
                        new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, 1000));
                System.out.printf("Reading tags with a TID manufacturer of 0x003\n",
                        filter.toString());
                filteredTagReads = r.read(500);
                for (TagReadData tr : filteredTagReads)
                {
                    System.out.println(tr.toString());
                }
                System.out.println();
                
                /*
                * A filter can also perform Gen2 Truncate Select operation. 
                * Truncate indicates whether a Tag’s backscattered reply shall be truncated to those EPC bits that follow Mask.
                * For example, truncated select starting with PC word start address 0x10 and length of 32 bits
                */
                if (r.getClass().getName() == ("com.thingmagic.SerialReader")) 
                {
                    filter = new Gen2.Select(false, Gen2.Bank.GEN2EPCTRUNCATE, 0x10, 0x20, new byte[]{(byte) 0x30, (byte) 0x00,
                        (byte) 0xAB, (byte) 0xCD});
                    r.paramSet("/reader/read/plan",
                            new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, 1000));
                    System.out.printf("Reading tags with Gen2 Truncate operation to those EPC's that follows the mask\n",
                            filter.toString());
                    filteredTagReads = r.read(500);
                    for (TagReadData tr : filteredTagReads) {
                        System.out.println(tr.toString());
                    }
                    System.out.println();

                    /*
                     * A filter can also perform Gen2 Tag Filtering.
                     * Major advantage of this feature is to limit the EPC response to user specified length field and all others will be rejected by firmware.
                     * invert, bitPointer, mask : Parameters will be ignored when GEN2EPCLENGTHFILTER is used
                     * maskBitLength : Specified EPC Length used for filtering
                     * For example, Tag filtering will be applied on EPC with 128 bits length, rest of the tags will be ignored
                     */
                    filter = new Gen2.Select(false, Gen2.Bank.GEN2EPCLENGTHFILTER, 16, 128, new byte[]{(byte) 0x30, (byte) 0x00});
                    r.paramSet("/reader/read/plan",
                            new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, 1000));
                    System.out.printf("Reading tags with Gen2 Tag EPC filtering operation \n",
                            filter.toString());
                    filteredTagReads = r.read(500);
                    for (TagReadData tr : filteredTagReads) {
                        System.out.println(tr.toString());
                    }
                    System.out.println();
                }
                // Filters can also be used to match tags that have already been
                // read. This form can only match on the EPC, as that's the only
                // data from the tag's memory that is contained in a TagData
                // object.
                // Note that this filter has invert=true. This filter will match
                // tags whose bits do not match the selection mask.
                // Also note the offset - the EPC code starts at bit 32 of the
                // EPC memory bank, after the StoredCRC and StoredPC.
                // In case of Network readers, ensure that bitLength is a multiple of 8.
                filter = new Gen2.Select(true, Gen2.Bank.EPC, 32, 2,new byte[] {(byte)0xC0});
                System.out.println("EPCs with first 2 bits equal to zero (post-filtered):");
                for (TagReadData tr : tagReads) // unfiltered tag reads from the first example
                {
                    if (filter.matches(tr.getTag()))
                    {
                        System.out.println(tr.toString());
                    }
                }
                System.out.println();

                {
                    /**
                     * Multi filter creation and initialization. This filter will match those tags whose bits matches the
                     * selection mask of both tidFilter(tags manufactured by Alien Technology) and epcFilter(epc of first
                     * tag read from tagReads[0]). Target and action are the two new parameters of Gen2.Select class whose
                     * default values could be "Gen2.Select.Target.Select" and "Gen2.Select.Action.ON_N_OFF" respectively
                     * if not provided by the user.
                     *
                     * Gen2 Select Action indicates which Select action to take
                     * (See Gen2 spec /Select commands / Tag response to Action parameter)
                     * |-----------------------------------------------------------------------------|
                     * |  Action  |        Tag Matching            |        Tag Not-Matching         |
                     * |----------|--------------------------------|---------------------------------|
                     * |   0x00   |  Assert SL or Inventoried->A   |  Deassert SL or Inventoried->B  |
                     * |   0x01   |  Assert SL or Inventoried->A   |  Do nothing                     |
                     * |   0x02   |  Do nothing                    |  Deassert SL or Inventoried->B  |
                     * |   0x03   |  Negate SL or (A->B,B->A)      |  Do nothing                     |
                     * |   0x04   |  Deassert SL or Inventoried->B |  Assert SL or Inventoried->A    |
                     * |   0x05   |  Deassert SL or Inventoried->B |  Do nothing                     |
                     * |   0x06   |  Do nothing                    |  Assert SL or Inventoried->A    |
                     * |   0x07   |  Do nothing                    |  Negate SL or (A->B,B->A)       |
                     * -------------------------------------------------------------------------------
                     *
                     * To improve readability and ease typing, these names abbreviate the official terminology of the Gen2 spec.
                     * <A>_N_<B>: The "_N_" stands for "Non-Matching".
                     * The <A> clause before the _N_ describes what happens to Matching tags.
                     * The <B> clause after the _N_ describes what happens to Non-Matching tags.
                     * (Alternately, you can pronounce "_N_" as "and", or "&"; i.e.,
                     * the pair of Matching / Non-Matching actions is known as "<A> and <B>".)
                     *
                     * ON: assert SL or inventoried -> A
                     * OFF: deassert SL or inventoried -> B
                     * NEG: negate SL or (A->B, B->A)
                     * NOP: do nothing
                     *
                     * The enum is simply a transliteration of the Gen2 spec's table: "Tag response to Action parameter"
                     */

                    // create and initialize tidFilter.
                    // In case of Network readers, ensure that bitLength is a multiple of 8.
                    Gen2.Select tidFilter = new Gen2.Select(false, Gen2.Bank.TID, 8, 12, new byte[] {(byte)0x00, (byte)0x30});
                    tidFilter.target = Gen2.Select.Target.Select;
                    tidFilter.action = Gen2.Select.Action.ON_N_OFF;

                    // create and initialize epcFilter
                    Gen2.Select epcFilter = new Gen2.Select(false, Gen2.Bank.EPC, 32, 16, ReaderUtil.hexStringToByteArray(tagReads[0].epcString()));
                    epcFilter.target = Gen2.Select.Target.Select;
                    epcFilter.action = Gen2.Select.Action.ON_N_OFF;

                    // Initialize multifilter with tagFilter array containing list of filters
                    MultiFilter multiFilter = new MultiFilter(new TagFilter[]{tidFilter, epcFilter});
                    r.paramSet("/reader/read/plan", new SimpleReadPlan(antennaList, TagProtocol.GEN2, multiFilter, 1000));
                    System.out.printf("Reading tags which matches multi filter criteria \n",filter.toString());
                    filteredTagReads = r.read(500);
                    for (TagReadData tr : filteredTagReads) 
                    {
                        System.out.println(tr.toString());
                    }
                    System.out.println();
                }
            }
      }
      finally
      {
        // Restore original settings
        System.out.println("Restoring Session " + oldSession);
        r.paramSet("/reader/gen2/session", oldSession);
      }

      // Shut down reader
      r.destroy();
    } 
    catch (ReaderException re)
    {
      System.out.println("ReaderException: " + re.getMessage());
    }
    catch (Exception re)
    {
        System.out.println("Exception: " + re.getMessage());
    }
  }
  
  static  int[] parseAntennaList(String[] args,int argPosition)
    {
        int[] antennaList = null;
        try
        {
            String argument = args[argPosition + 1];
            String[] antennas = argument.split(",");
            int i = 0;
            antennaList = new int[antennas.length];
            for (String ant : antennas)
            {
                antennaList[i] = Integer.parseInt(ant);
                i++;
            }
        }
        catch (IndexOutOfBoundsException ex)
        {
            System.out.println("Missing argument after " + args[argPosition]);
            usage();
        }
        catch (Exception ex)
        {
            System.out.println("Invalid argument at position " + (argPosition + 1) + ". " + ex.getMessage());
            usage();
        }
        return antennaList;
    }
}

File: MercuryAPI-1.31.3.36/c/src/samples/filter.c
/**
 * Sample program that demonstrates different types and uses of TagFilter objects.
 * @file filter.c
 */

#include <tm_reader.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <tmr_utils.h>
#include <string.h>
#include <inttypes.h>

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif

#define usage() {errx(1, "Please provide valid reader URL, such as: reader-uri [--ant n]\n"\
                         "reader-uri : e.g., 'tmr:///COM1' or 'tmr:///dev/ttyS0/' or 'tmr://readerIP'\n"\
                         "[--ant n] : e.g., '--ant 1'\n"\
                         "Example: 'tmr:///com4' or 'tmr:///com4 --ant 1,2' \n");}

void errx(int exitval, const char *fmt, ...)
{
  va_list ap;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);

  exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
  if (TMR_SUCCESS != ret)
  {
    errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
  }
}

void readAndPrintTags(TMR_Reader *rp, int timeout)
{
  TMR_Status ret;
  TMR_TagReadData trd;
  char epcString[128];

  ret = TMR_read(rp, timeout, NULL);
  if (TMR_ERROR_TAG_ID_BUFFER_FULL == ret)
  {
    /* In case of TAG ID Buffer Full, extract the tags present
    * in buffer.
    */
    fprintf(stdout, "reading tags:%s\n", TMR_strerr(rp, ret));
  }
  else
  {
    checkerr(rp, ret, 1, "reading tags");
  }

  while (TMR_SUCCESS == TMR_hasMoreTags(rp))
  {
    ret = TMR_getNextTag(rp, &trd);
    checkerr(rp, ret, 1, "fetching tag");
    TMR_bytesToHex(trd.tag.epc, trd.tag.epcByteCount, epcString);
    printf("EPC : %s\n", epcString);
  }
}  

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
                   uint32_t timeout, void *cookie)
{
  FILE *out = cookie;
  uint32_t i;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  for (i = 0; i < dataLen; i++)
  {
    if (i > 0 && (i & 15) == 0)
    {
      fprintf(out, "\n         ");
    }
    fprintf(out, " %02x", data[i]);
  }
  fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
  FILE *out = cookie;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  fprintf(out, "%s\n", data);
}

void parseAntennaList(uint8_t *antenna, uint8_t *antennaCount, char *args)
{
  char *token = NULL;
  char *str = ",";
  uint8_t i = 0x00;
  int scans;

  /* get the first token */
  if (NULL == args)
  {
    fprintf(stdout, "Missing argument\n");
    usage();
  }

  token = strtok(args, str);
  if (NULL == token)
  {
    fprintf(stdout, "Missing argument after %s\n", args);
    usage();
  }

  while(NULL != token)
  {
    scans = sscanf(token, "%"SCNu8, &antenna[i]);
    if (1 != scans)
    {
      fprintf(stdout, "Can't parse '%s' as an 8-bit unsigned integer value\n", token);
      usage();
    }
    i++;
    token = strtok(NULL, str);
  }
  *antennaCount = i;
}

int main(int argc, char *argv[])
{
  TMR_Reader r, *rp;
  TMR_Status ret;
  TMR_Region region;
  TMR_TagReadData trd;
  TMR_TagFilter filter;
  TMR_ReadPlan filteredReadPlan;
  TMR_TagOp tagop;
  char epcString[128];
  uint8_t mask[2];
  uint8_t tmask[4];
  uint8_t *antennaList = NULL;
  uint8_t buffer[20];
  uint8_t i;
  uint8_t antennaCount = 0x0;
#if USE_TRANSPORT_LISTENER
  TMR_TransportListenerBlock tb;
#endif

  if (argc < 2)
  {
    usage();
  }

  for (i = 2; i < argc; i+=2)
  {
    if(0x00 == strcmp("--ant", argv[i]))
    {
      if (NULL != antennaList)
      {
        fprintf(stdout, "Duplicate argument: --ant specified more than once\n");
        usage();
      }
      parseAntennaList(buffer, &antennaCount, argv[i+1]);
      antennaList = buffer;
    }
    else
    {
      fprintf(stdout, "Argument %s is not recognized\n", argv[i]);
      usage();
    }
  }  
  
  rp = &r;
  
  // Create Reader object, connecting to physical device
  ret = TMR_create(rp, argv[1]);
  checkerr(rp, ret, 1, "creating reader");

#if USE_TRANSPORT_LISTENER

  if (TMR_READER_TYPE_SERIAL == rp->readerType)
  {
    tb.listener = serialPrinter;
  }
  else
  {
    tb.listener = stringPrinter;
  }
  tb.cookie = stdout;

  TMR_addTransportListener(rp, &tb);
#endif

  ret = TMR_connect(rp);
  checkerr(rp, ret, 1, "connecting reader");

  region = TMR_REGION_NONE;
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
  checkerr(rp, ret, 1, "getting region");

  if (TMR_REGION_NONE == region)
  {
    TMR_RegionList regions;
    TMR_Region _regionStore[32];
    regions.list = _regionStore;
    regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
    regions.len = 0;

    ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
    checkerr(rp, ret, __LINE__, "getting supported regions");

    if (regions.len < 1)
    {
      checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't supportany regions");
    }
    region = regions.list[0];
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
    checkerr(rp, ret, 1, "setting region");  
  }

  /**
   * Checking the software version of the sargas.
   * The antenna detection is supported on sargas from software version of 5.3.x.x.
   * If the Sargas software version is 5.1.x.x then antenna detection is not supported.
   * User has to pass the antenna as arguments
   */
  {
    ret = isAntDetectEnabled(rp, antennaList);
    if(TMR_ERROR_UNSUPPORTED == ret)
    {
      fprintf(stdout, "Reader doesn't support antenna detection. Please provide antenna list.\n");
      usage();
    }
    else
    {
      checkerr(rp, ret, 1, "Getting Antenna Detection Flag Status");
    }
  }
  // initialize the read plan
  ret = TMR_RP_init_simple(&filteredReadPlan, antennaCount, antennaList, TMR_TAG_PROTOCOL_GEN2, 1000);
  checkerr(rp, ret, 1, "initializing the  read plan");
  ret = TMR_paramSet(rp, TMR_paramID("/reader/read/plan"), &filteredReadPlan);
  checkerr(rp, ret, 1, "setting read plan");

  ret = TMR_read(rp, 500, NULL);
  if (TMR_ERROR_TAG_ID_BUFFER_FULL == ret)
  {
    /* In case of TAG ID Buffer Full, extract the tags present
    * in buffer.
    */
    fprintf(stdout, "reading tags:%s\n", TMR_strerr(rp, ret));
  }
  else
  {
    checkerr(rp, ret, 1, "reading tags");
  }

  if (TMR_ERROR_NO_TAGS == TMR_hasMoreTags(rp))
  {
    errx(1, "No tags found for test\n");
  }

  ret = TMR_getNextTag(rp, &trd);
  checkerr(rp, ret, 1, "getting tags");

  /*
   * A TagData object may be used as a filter, for example to
   * perform a tag data operation on a particular tag.
   * Read kill password of tag found in previous operation
   */
   TMR_TF_init_tag(&filter, &trd.tag);
   TMR_bytesToHex(filter.u.tagData.epc, filter.u.tagData.epcByteCount,
                 epcString);
   printf("Read kill password of tag that begin with %s\n",epcString);
   TMR_TagOp_init_GEN2_ReadData(&tagop,TMR_GEN2_BANK_RESERVED,0,2);
   TMR_executeTagOp(rp,&tagop,&filter,NULL);

  /*
   * Filter objects that apply to multiple tags are most useful in
   * narrowing the set of tags that will be read. This is
   * performed by setting a read plan that contains a filter.
   */
  TMR_TF_init_tag(&filter, &trd.tag);
  TMR_bytesToHex(filter.u.tagData.epc, filter.u.tagData.epcByteCount,
                 epcString);
  printf("Reading tags that begin with %s\n", epcString);
  readAndPrintTags(rp, 500);

      
  /*
   * A TagData with a short EPC will filter for tags whose EPC
   * starts with the same sequence.
   */
  filter.type = TMR_FILTER_TYPE_TAG_DATA;
  filter.u.tagData.epcByteCount = 4;
  tm_memcpy(filter.u.tagData.epc, trd.tag.epc, (size_t)filter.u.tagData.epcByteCount);
  
  TMR_RP_init_simple(&filteredReadPlan,
                     antennaCount, antennaList, TMR_TAG_PROTOCOL_GEN2, 1000);
  TMR_RP_set_filter(&filteredReadPlan, &filter);

  ret = TMR_paramSet(rp, TMR_paramID("/reader/read/plan"), &filteredReadPlan);
  checkerr(rp, ret, 1, "setting read plan");

  TMR_bytesToHex(filter.u.tagData.epc, filter.u.tagData.epcByteCount,
                 epcString);
  printf("Reading tags that begin with %s\n", epcString);
  readAndPrintTags(rp, 500);

  /*
   * A filter can also be an explicit Gen2 Select operation.  For
   * example, this filter matches all Gen2 tags where bits 8-19 of
   * the TID are 0x30 (that is, tags manufactured by Alien
   * Technology).
   */
  // In case of Network readers, ensure that bitLength is a multiple of 8
  mask[0] = 0x00;
  mask[1] = 0x30;
  TMR_TF_init_gen2_select(&filter, false, TMR_GEN2_BANK_TID, 8, 12, mask);
  /*
   * filteredReadPlan already points to filter, and
   * "/reader/read/plan" already points to filteredReadPlan.
   * However, we need to set it again in case the reader has 
   * saved internal state based on the read plan.
   */
  ret = TMR_paramSet(rp, TMR_paramID("/reader/read/plan"), &filteredReadPlan);
  checkerr(rp, ret, 1, "setting read plan");
  printf("Reading tags with a TID manufacturer of 0x0030\n");
  readAndPrintTags(rp, 500);

	if (rp->readerType == TMR_READER_TYPE_SERIAL)
  {
		/*
		 * A filter can also perform Gen2 Truncate Select operation. 
		 * Truncate indicates whether a Tag\92s backscattered reply shall be truncated to those EPC bits that follow Mask.
		 * For example, truncated select starting with PC word start address and length of 16 bits
		 */
		// In case of Network readers, ensure that bitLength is a multiple of 8
		tmask[0] = 0x30;
		tmask[1] = 0x00;
		tmask[2] = 0xDE;
		tmask[3] = 0xAD;
		TMR_TF_init_gen2_select(&filter, false, TMR_GEN2_EPC_TRUNCATE, 16, 32, tmask);
		/*
		 * filteredReadPlan already points to filter, and
		 * "/reader/read/plan" already points to filteredReadPlan.
		 * However, we need to set it again in case the reader has 
		 * saved internal state based on the read plan.
		 */
		ret = TMR_paramSet(rp, TMR_paramID("/reader/read/plan"), &filteredReadPlan);
		checkerr(rp, ret, 1, "setting read plan");
		printf("Gen2 Truncate operation to those EPC that follows the mask\n");
		readAndPrintTags(rp, 500);

		/*
		 * A filter can also perform Gen2 Tag Filtering.
		 * Major advantage of this feature is to limit the EPC response to user specified length field and all others will be rejected by firmware.
		 * invert, bitPointer, mask : Parameters will be ignored when TMR_GEN2_EPC_LENGTH_FILTER is used
		 * maskBitLength : Specified EPC Length used for filtering
		 * For example, Tag filtering will be applied on EPC with 128 bits length, rest of the tags will be ignored
		 */
		// In case of Network readers, ensure that bitLength is a multiple of 8
		mask[0] = 0x30;
		mask[1] = 0x00;
		TMR_TF_init_gen2_select(&filter, false, TMR_GEN2_EPC_LENGTH_FILTER, 16, 128, mask);
		/*
		 * filteredReadPlan already points to filter, and
		 * "/reader/read/plan" already points to filteredReadPlan.
		 * However, we need to set it again in case the reader has 
		 * saved internal state based on the read plan.
		 */
		ret = TMR_paramSet(rp, TMR_paramID("/reader/read/plan"), &filteredReadPlan);
		checkerr(rp, ret, 1, "setting read plan");
		printf("Filtering Tag with EPC Length of %d bits\n", filter.u.gen2Select.maskBitLength);
		readAndPrintTags(rp, 500);
	}

  /*
   * Filters can also be used to match tags that have already been
   * read. This form can only match on the EPC, as that's the only
   * data from the tag's memory that is contained in a TagData
   * object.
   * Note that this filter has invert=true. This filter will match
   * tags whose bits do not match the selection mask.
   * Also note the offset - the EPC code starts at bit 32 of the
   * EPC memory bank, after the StoredCRC and StoredPC.
   */
  // In case of Network readers, ensure that bitLength is a multiple of 8
  TMR_TF_init_gen2_select(&filter, true, TMR_GEN2_BANK_EPC, 32, 2, mask);

  printf("Reading tags with EPC's having first two bytes equal to zero (post-filtered):\n");
  ret = TMR_read(rp, 500, NULL);
  if (TMR_ERROR_TAG_ID_BUFFER_FULL == ret)
  {
    /* In case of TAG ID Buffer Full, extract the tags present
    * in buffer.
    */
    fprintf(stdout, "reading tags:%s\n", TMR_strerr(rp, ret));
  }
  else
  {
    checkerr(rp, ret, 1, "reading tags");
  }

  while (TMR_SUCCESS == TMR_hasMoreTags(rp))
  {
    ret = TMR_getNextTag(rp, &trd);
    checkerr(rp, ret, 1, "fetching tag");
    if (TMR_TF_match(&filter, &trd.tag))
    {
      TMR_bytesToHex(trd.tag.epc, trd.tag.epcByteCount, epcString);
      printf("%s\n", epcString);
    }
  }
  {
    /**
     * Multi filter creation and initialization. This filter will match those tags whose bits matches the
     * selection mask of both tidFilter(tags manufactured by Alien Technology) and epcFilter.
     * Target and action are the two new parameters of TMR_GEN2_Select structure whose
     * default values could be "gen2Select.target =  SELECT" and "gen2Select.action = ON_N_OFF" respectively
     * if not provided by the user.
     *******************************************************************************
     TMR_GEN2_Select_action indicates which Select action to take
     (See Gen2 spec /Select commands / Tag response to Action parameter)
     |-----------------------------------------------------------------------------|
     |  Action  |        Tag Matching            |        Tag Not-Matching         |
     |----------|--------------------------------|---------------------------------|
     |   0x00   |  Assert SL or Inventoried->A   |  Deassert SL or Inventoried->B  |
     |   0x01   |  Assert SL or Inventoried->A   |  Do nothing                     |
     |   0x02   |  Do nothing                    |  Deassert SL or Inventoried->B  |
     |   0x03   |  Negate SL or (A->B,B->A)      |  Do nothing                     |
     |   0x04   |  Deassert SL or Inventoried->B |  Assert SL or Inventoried->A    |
     |   0x05   |  Deassert SL or Inventoried->B |  Do nothing                     |
     |   0x06   |  Do nothing                    |  Assert SL or Inventoried->A    |
     |   0x07   |  Do nothing                    |  Negate SL or (A->B,B->A)       |
     ********************************************************************************
     *
     *  To improve readability and ease of typing, these names abbreviate the official terminology of the Gen2 spec.
     *  <A>_N_<B>: The "_N_" stands for "Non-Matching".
     *  The <A> clause before the N describes what happens to Matching tags.
     *  The <B> clause after the N describes what happens to Non-Matching tags.
     *  (Alternately, you can pronounce "_N_" as "and", or "&"; i.e.,
     *  the pair of Matching / Non-Matching actions is known as "<A> and <B>".)
     *
     *  ON: assert SL or inventoried -> A
     *  OFF: deassert SL or inventoried -> B
     *  NEG: negate SL or (A->B, B->A)
     *  NOP: do nothing
     *
     *  The enum is simply a transliteration of the Gen2 spec's table: "Tag response to Action parameter"
     */
     TMR_TagFilter tidFilter, epcFilter, *filterArray[2];
     TMR_MultiFilter filterList;
 
     filterList.tagFilterList = filterArray;
     filterList.len = 0;
     {
       /* This select filter matches all Gen2 tags where bits 08-12 of the TID are 0x0030(that is, tags manufactured
        * by Alien Technology). */ 
	   // In case of Network readers, ensure that bitLength is a multiple of 8
       mask[0] = 0x00;
       mask[1] = 0x30;
       TMR_TF_init_gen2_select(&tidFilter, false, TMR_GEN2_BANK_TID, 8, 12, mask);
       tidFilter.u.gen2Select.target =  SELECT;
       tidFilter.u.gen2Select.action =  ON_N_OFF;
     }
     {
       /* This select filter matches all Gen2 tags whose EPC starts with the same sequence */
	   // In case of Network readers, ensure that bitLength is a multiple of 8.
       TMR_TF_init_gen2_select(&epcFilter, false, TMR_GEN2_BANK_EPC, 32, 16, trd.tag.epc);
       epcFilter.u.gen2Select.target =  SELECT;
       epcFilter.u.gen2Select.action =  NOP_N_OFF;
     }
     /*Assemble two filters in filterArray*/
     filterArray[filterList.len++] = &tidFilter;
     filterArray[filterList.len++] = &epcFilter;

     /* Assign TMR_GEN2_MultiSelect filter to TMR_TagFilter filter */
     filter.type = TMR_FILTER_TYPE_MULTI;
     filter.u.multiFilterList = filterList;
     printf("Reading tags with double select \n");
     readAndPrintTags(rp, 500);
  }

  TMR_destroy(rp);
  return 0;
}
